aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -571,95 +571,105 @@ Tensor sparse_compressed_to_dense(
+      dense_reshaped_sizes[1] = blocksize[0];
+      dense_reshaped_sizes[2] = blocksize[1];
     }
-    // Due to early exit above this reshape should always be valid
-    dense = dense.reshape({n_batch, -1, values.size(-2), values.size(-1)});
-    for (auto batch : c10::irange(n_batch)) {
-      Tensor batch_indices = at::_convert_indices_from_csr_to_coo(
-          compressed_indices[batch],
-          plain_indices[batch],
-          false,
-          self.layout() == kSparseBsc);
-      auto batch_row_indices = batch_indices.select(0, 0);
-      auto batch_col_indices = batch_indices.select(0, 1);
-      auto offsets = batch_col_indices +
-          batch_row_indices * (self.size(-1) / blocksize[1]);
-      dense[batch].index_add_(0, offsets, values[batch]);
+    dense_reshaped_sizes[0] = n_batch * nrows * ncols;
+    dense = dense.reshape(dense_reshaped_sizes);
+
+    // Calculate batch, row and column indices for non-zeros in the
+    // sparse matrix, and use these to calculate correspoding indices
+    // into the dense matrix reshaped as above.  Then, update dense
+    // matrix by adding sparse matrix values into elements with
+    // indices calculated this way.
+    auto options = compressed_indices.options();
+    auto nnz_per_batch = values.size(1);
+    auto batch_indices = at::arange(0, n_batch, options).repeat_interleave(nnz_per_batch);
+    auto ncompressed = compressed_rows ? nrows : ncols;
+    auto compressed_indices_over_all_batches =
+      at::cat({compressed_indices.slice(1, 0, ncompressed).flatten()
+              + nnz_per_batch * at::arange(0, n_batch, options).repeat_interleave(ncompressed),
+              n_batch * nnz_per_batch * at::ones({1}, options)});
+    Tensor indices = at::_convert_indices_from_csr_to_coo(
+        compressed_indices_over_all_batches,
+        plain_indices.flatten(),
+        false,
+        !compressed_rows);
+    auto row_indices = indices.select(0, 0);
+    auto col_indices = indices.select(0, 1);
+    if (compressed_rows) {
+      row_indices -= batch_indices * nrows;
+    } else {
+      col_indices -= batch_indices * ncols;
     }
+    auto offsets = col_indices + row_indices * ncols + batch_indices * nrows * ncols;
+    dense.index_add_(0, offsets, values.flatten(0, 1));
 
-    // un-tile the result, NOTE: The final reshape uses the original
-    // self.sizes() which will squeeze out the extra batch dim if we put one in
-    return dense
-        .unflatten(
-            1, {self.size(-2) / blocksize[0], self.size(-1) / blocksize[1]})
+    // Un-tile the result.  The final reshape uses the original
+    // self.sizes() which will squeeze out the extra batch dim if we
+    // put one in.
+    if (!block_sparse) {
+      return dense.reshape(self.sizes());
+    } else {
+      return dense
+        .unflatten(0, {-1, nrows, ncols})
         .transpose(2, 3)
         .reshape(self.sizes());
+    }
   }
+
   return self.to_sparse().to_dense();
```

Comment:
Continued:
```suggestion
```

Reply:
Applied all suggestions, except for `numel()`.  Namely, if attached [patch.txt](https://github.com/pytorch/pytorch/files/10362654/patch.txt) applied say to [6a83954](https://github.com/pytorch/pytorch/commit/6a83954b1a52c1c22520c236d99a1cff879173a5), then for example `test_sampled_addmm_cuda_float32` from `test/test_sparse_csr.py` will fail, on qgpu3, printing out following:
```python
actual         =  tensor(crow_indices=tensor([0, 2, 3]),
       col_indices=tensor([0, 1, 0]),
       values=tensor([5.4367, 2.6152, 0.5298]), device='cuda:0', size=(2, 2),
       nnz=3, layout=torch.sparse_csr)
actual.numel() =  0
```
Thus, `numel()` returns 0 for tensor that actually has non-zero elements.  Seems like an issue in `numel()`, however it can't be reproduced by simple:
```python
csr = torch.sparse_csr_tensor(crow_indices=torch.tensor([0, 2, 3]),
                              col_indices=torch.tensor([0, 1, 0]),
                              values=torch.tensor([5.4367, 2.6152, 0.5298]),
                              size=(2, 2),
                              device='cuda')
print(csr.numel())
```

Reply:
This is weird, locally when using ca 1 week old master, it works for me:
```python
>>> csr = torch.sparse_csr_tensor(crow_indices=torch.tensor([0, 2, 3]), col_indices=torch.tensor([0, 1, 0]), values=torch.tensor([5.4367, 2.6152, 0.5298]), size=(2, 2), device='cuda')
>>> csr.numel()
4
```

Reply:
This will reproduce it, with 2555971 (that is recent HEAD of viable/strict):
```python
dtype = torch.float32
device = 'cpu'
a = torch.ones((2, 2), dtype=dtype, device=device)
b = torch.ones((2, 2), dtype=dtype, device=device)
c = torch.sparse_csr_tensor(crow_indices=torch.tensor([0, 2, 3]),
                            col_indices=torch.tensor([0, 1, 0]),
                            values=torch.tensor([1, 1, 1]),
                            size=(2, 2),
                            dtype=dtype,
                            device=device,)
actual = torch.sparse.sampled_addmm(c, a, b, alpha=1, beta=1)
print(actual)
print(actual.numel())
```

Reply:
@alexsamardzic could you create an issue about this?

Reply:
I believe it is a bug in `resize_`:
```python
>>> torch.empty((0, 0), layout=torch.sparse_csr).resize_((2, 2)).numel()
0
```

Reply:
I have created https://github.com/pytorch/pytorch/issues/91830 .